handlers/proxy_handler.cc

```diff
@@ -0,0 +1,240 @@
+	return SendRequestToServer(m_host_path_, m_port_path_, request, response);
+}
+
+// convert a local uri like "/proxy/some/path" to remote uri "/some/path"
+std::string ProxyHandler::ExtractNonProxyUri(const std::string& prefix, const std::string& uri) {
+	if (uri.size() < prefix.size()) {
+		printf("ProxyHandler: warning: request uri is shorter than uri prefix!\n");
+		return uri;
+	}
+
+	std::string modified_uri = uri.substr(prefix.size());
+	if (modified_uri == "") modified_uri = "/";
+	if (modified_uri[0] != '/') modified_uri = "/" + modified_uri;
+	return modified_uri;
+}
+
+void ProxyHandler::ModifyRequestForProxy(const Request& request, MutableRequest* modified_request) {
+	std::string modified_uri = ExtractNonProxyUri(m_uri_prefix_, request.uri());
+	printf("ProxyHandler: requesting '%s' ('%s')\n", modified_uri.c_str(), request.uri().c_str());
+
+	modified_request->SetURI(modified_uri);
+	modified_request->SetHeader("Connection", "close");
+	modified_request->SetHeader("Host", m_host_path_ + ":" + m_port_path_);
+}
+
+// given a location string like "http://www.ucla.edu/some/path", extract "www.ucla.edu" and "/some/path"
+void ProxyHandler::ParseRedirectLocation(std::string location, std::string* new_path, std::string* new_host) {
+	location = boost::algorithm::trim_copy(location);
+
+	// strip "http://", if present
+	const size_t http_size = 7; // "http://"
+	if (location.size() > http_size && location.substr(0, http_size) == "http://") {
+		location = location.substr(http_size);
+	}
+
+	const size_t https_size = 8; // "https://"
+	if (location.size() > https_size && location.substr(0, https_size) == "https://") {
+		printf("ProxyHandler: warning: redirect specified https, which is unsupported.\n");
+		location = location.substr(https_size);
+	}
+
+	// Location will be in form `newhost/new/path`, we want to split into 'newhost' + '/new/path'
+	auto pathPos = location.find('/');
+	if (pathPos != std::string::npos) {
+		*new_host = location.substr(0, pathPos);
+		*new_path = location.substr(pathPos);
+	} else {
+		*new_host = location;
+		*new_path = "/"; // no path given
+	}
+}
+
+void ProxyHandler::ConnectSocketToEndpoint(tcp::socket* socket, std::string host, std::string port) {
+	tcp::resolver resolver(io_service_);
+	tcp::resolver::query query(host, port);
+	tcp::resolver::iterator endpoint_iterator = resolver.resolve(query);
+	boost::asio::connect(*socket, endpoint_iterator);
+}
+
+void ProxyHandler::WriteToSocket(tcp::socket* socket, const std::string& requestStr) {
+	boost::asio::streambuf reqBuf;
+	std::ostream request_stream(&reqBuf);
+	request_stream << requestStr;
+	boost::asio::write(*socket, reqBuf);
+}
+
+boost::system::error_code ProxyHandler::SocketReadToEOF(tcp::socket* socket, boost::asio::streambuf* buf, std::string* data) {
+	boost::system::error_code error;
+	while (true) {
+		boost::asio::read(*socket, *buf, boost::asio::transfer_at_least(1), error);
+		if (buf->size() == 0) {
```

Comment:
In the event of an error, such as EOF, it can read 0 I believe. This is also just defensive, in case semantics of how we choose to transfer change later, a future developer can change just the `read` call. It may be the case that the proxy'd server returns an empty body.